### DISCUSSION TITLE
Upgrading from `convert` to `magick` messes up PNG options

### DISCUSSION THREAD

**Discussion by ilyagr** [ImageMagick/ImageMagick#7596] [Help] [answered]
### ImageMagick version

7.1.1-37

### Operating system, version and so on

Mac OS X Sonoma on ARM
I think ImageMagick might be using Inkscape to do the SVG->PNG conversion.

### Description

I am using ImageMagick to convert SVG to PNG. Previously, I used the following command:

     convert  -colors 63 -background black -resize 100%  \
            "$script_base".svg "$script_base".png

After  moving to a different computer and upgrading ImageMagick, that command started generating warnings. So, I [changed the command](https://github.com/martinvonz/jj/commit/6d573ef6d7a45151495de18b6f4c5063ce39f6bd) to

    magick "$script_base".svg \
            -colors 63 -background black -resize 100%  \
            "$script_base".png

This *seems* to work, but is incorrect on a more careful look. Neither `-background black` nor `-colors 63` seem to have the desired effect anymore.  Here is the [comparison of before-and-after](https://github.com/martinvonz/jj/commit/42dee7d08ce8e362cf9d44f844b25e001b6ac94f). Note how, in the new version,

- The corners are white, not black (if you zoom in a bit). The original SVGs have transparent background that shows in these corners.
- The resulting PNGs are much larger. Checking them with an image viewer, the new PNGs are RGB-color not 8-bit indexed color like previous ones.

I am pretty sure the changes in the SVG files are irrelevant, see [here](https://github.com/martinvonz/jj/commit/1cc10783a2fcd6e0261a1dd9cfee28a05a838f64#diff-64a71bf99d69a916fc5bc93205c9cf661fe73b5ceabc6b0e01f7bc9397c1a636) for an example of how they changed.

Is there a way to fix this?

P.S. I'm currently running Mac OS X on ARM

```version
$ magick --version
Version: ImageMagick 7.1.1-37 Q16-HDRI aarch64 5b2825b29:20240824 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI
Delegates (built-in): bzlib cairo djvu fontconfig freetype heic jng jp2 jpeg jxl lcms lqr lzma openexr pangocairo png raw rsvg tiff webp x xml zlib zstd
Compiler: gcc (4.2)
```

**Comment by fmw42** [comment]
Syntax in IM 7 is more strict than in IM 6.  Both properly do not use normal Unix convention.  In Imagemagick, you typically read a raster input right after magick.  See https://imagemagick.org/Usage/basics/#cmdline

So

`magick  "$script_base".svg -colors 63 -background black -resize 100%  "$script_base".png`

Why would you resize by 100%?  That is full resolution. And your input is already full resolution. Perhaps I misunderstand.

If you need to change the size of the output, I would suggest putting -density XX before the input, since you have a vector SVG file.  Nominal density for SVG I believe is 96 dpi, which is full resolution. If you want something larger, then make the density larger than 96.

**Reply by ilyagr** [reply]
Thanks for looking!

I don't care about the `-resize`. It is indeed a no-op, I should have removed it from the question. I care about making the transparent corners of the SVG black (as opposed to white) and reducing the size of the PNG by reducing the number of colors.

**Reply by fmw42** [reply]
Then try

`magick  "$script_base".svg -colors 63 -background black -flatten  "$script_base".png`

Does that do what you want?

**Reply by ilyagr** [reply]
No, this produces seemingly exactly the same result as

```
magick "$script_base".svg \
        -colors 63 -background black -resize 100%  \
        "$script_base".png
```

Unlike the convert command, the corners are still white and the file is large.

Actually, after some searching through the docs, adding `-depth 8` to the command seems to help somewhat.


```
magick "$script_base".svg \
        -colors 63 -background black -resize 100% -depth 8 \
        "$script_base".png
```

produces a smaller file. But the corners are still the wrong color.

**Reply by fmw42** [reply]
Can you post your input SVG file for us to use to test IM 6 vs IM 7?  What renderer are you using:  Internal MSVG, RSVG delegate, Inkscape?  Inkscape is the best of the 3.

**Reply by fmw42** [reply]
Try

`magick  "$script_base".svg -colors 63 -background black -alpha background -alpha off  "$script_base".png`

Otherwise I would need to have your input SVG file to use to see what it has?  Are the transparent pixels you want to change fully transparent or partially transparent?

**Reply by ilyagr** [reply]
Thanks, I think I figured out what was going on more or less, thought it led to more questions. See https://github.com/ImageMagick/ImageMagick/discussions/7596#discussioncomment-10575050.

**Comment by snibgo** [comment, marked as answer]
The OP command is:
```
magick "$script_base".svg \
        -colors 63 -background black -resize 100%  \
        "$script_base".png
```
In that command, both `-background black` and `-resize 100%` have no effect on the output image. If you move `-background black` to *before* the input SVG, I expect the transparent parts of the rasterized SVG will black instead of white.

If that doesn't help, please post the SVG.

**Reply by ilyagr** [reply]
Thank you, that does help with one of the problems.

I find this very counter-intuitive. I would expect

    magick qq.svg qq.png

to preserve transparency. Then, it'd make sense for `magick qq.svg -background black qq.png` to make the background black. However, `magick qq.svg qq.png` actually turns all transparent pixels white, even though it's using Inkscape which is perfectly capable of exporting transparent PNGs.

Is this a bug, or worth a feature request?

In particular, while I requested a black background as a compromise, I would in fact prefer a comamnd that kept the background transparent. Is there a way to do that with ImageMagick?

-------------------

I went down a rabbit hole and I've figured out my other issue as well. It's also very un-intuitive, please let me know if this should be a bug or an FR. I'm also pretty sure the behavior changed in a recent version of ImageMagick even if I use the obsolete `magick convert`.

The issue is that the following two commands result in images that *look* identical, but the second one results in a much smaller image. For some reason, the first command can't figure out that it doesn't need full-color RGB.

    magick qq.svg -colors 63 qq.png
    magick qq.svg -colors 63 -depth 8 qq.png

There might be an even more efficient command, since I am not sure ImageMagick properly comes up with 63 indexed colors even in the latter case (do you know what that command would be?)

---------------

So, the **correct command** for what I'm doing seems to be

```
magick  -background black "$script_base".svg \
        -colors 63 -depth 8  \
        "$script_base".png
```

What a crazy upgrading story.

## Demo

Here are some examples, all operating on this SVG (which I hope GitHub represents faithfully).

![test](https://github.com/user-attachments/assets/1ce477ee-c9a1-4b19-9d65-f0caae1c71f3)

    magick test.svg out.png

![noop](https://github.com/user-attachments/assets/db4d45c4-9676-45a8-a79d-2548b6def7fd)

By running with `-verbose -debug Command`, I can tell that it invokes inkscape with

```
'inkscape' '/tmp/file' '--export-filename=/tmp/file.png' '--export-dpi=96' '--export-background=rgb(100%,100%,100%)' '--export-background-opacity=1'
```

Without the latter options, Inkscape can generate a perfectly fine PNG with transparency:

![inkscape-export](https://github.com/user-attachments/assets/a08d1f21-240b-4e0f-94e4-5398312d8549)


Finally, the following two files look identical, but the latter is half the size. You can see the difference in the RGB depth with `magick identify`.

```console
$ magick test.svg -colors 63 out.png
$ magick identify out.png  # This file is 7465 bytes, 16 bit
out.png PNG 192x192 192x192+0+0 16-bit sRGB 7465B 0.000u 0:00.003
```

![bgblcol63](https://github.com/user-attachments/assets/b2f8584e-de55-4d72-b073-d83cf7964db2)

```console
$ magick test.svg -colors 63 -depth 8 out.png
$ magick identify out.png  # This file is 2820 bytes, 8 bit
out.png PNG 192x192 192x192+0+0 8-bit sRGB 28c 2820B 0.000u 0:00.000
```

![bgblcol63dp8](https://github.com/user-attachments/assets/253d6b4d-0ea9-4206-9237-dc88545e22e0)

And finally, my precious correct command (though I'd actually prefer the same thing with transparency instead of black, can ImageMagick handle that?):

    magick -background black test.svg -colors 63 -depth 8 out.png

![bgblfirstcol63dp8](https://github.com/user-attachments/assets/d676fc21-f996-4879-b358-314c028d406d)

**Reply by ilyagr** [reply]
If the SVG uploaded to the comment doesn't work, you can grab another one in https://github.com/martinvonz/jj/blob/1cc10783a2fcd6e0261a1dd9cfee28a05a838f64/demos/juggle_conflicts.svg. There, however, the transparent rounded corners are much less visible unless you zoom in.

**Comment by snibgo** [comment]
>  I would in fact prefer a comamnd that kept the background transparent. 

Use `-background transparent` (or `-background none` which means the same) before reading the SVG.

> The issue is that the following two commands result in images that look identical, but the second one results in a much smaller image. For some reason, the first command can't figure out that it doesn't need full-color RGB.
magick qq.svg -colors 63 qq.png
magick qq.svg -colors 63 -depth 8 qq.png

"-colors 63" will find up to 63 colors that closely represent the input image. Unless your IM is Q8 non-HDMI, the color values are unlikely to have a precision of only 8 bits. For example, if your IM is Q16 non-HDMI, the colors probably need 16 bits per channel.

When you save an image at "-depth 8", the colors are further simplified to be no more than 8 bits per channel. This will reduce the file size in bytes, which perhaps is what you mean by "a much smaller image".

> What a crazy upgrading story.

When upgrading from v6 to v7, I suggest reading https://www.imagemagick.org/script/porting.php

**Reply by fmw42** [reply]
IM command syntax does not follow Unix command syntax. Especially IM 7.

See also https://imagemagick.org/Usage/basics/#cmdline

**Reply by ilyagr** [reply]
> Use -background transparent (or -background none which means the same) before reading the SVG.

This works, thanks, very good to know!

For other readers, the command would be

```
magick  -background none "$script_base".svg \
        -colors 63 -depth 8  \
        "$script_base".png
```

I have no idea how I could have guessed this without your help, especially since putting it after the input file doesn't work. 

> This will reduce the file size in bytes, which perhaps is what you mean by "a much smaller image".

Yes, that's what I meant. After some more running around the docs, I also found `-type Palette` which is what I actually wanted instead of `-depth 8` (I think? The results are very similar to `-depth 8` in terms of size, but asking for indexed colors seems like it would be better in theory).

> When upgrading from v6 to v7, I suggest reading https://www.imagemagick.org/script/porting.php

Thanks, I didn't find that page before. Though, looking it over, I found it difficult to find the sections relevant to my problems.

-------------------

Anyway, I think my problem is solved, thank you!

Here are two wishes that would have made my life much easier in retrospect:

- Make `-background none` the default when reading SVG files. This would also mean that `-background black` would work the same before or after reading the file.
- Make `-colors 63` imply something like `-type Palette`, to make compressing PNGs easier. I think this is similar to what an older ImageMagick did. This would also avoid another footgun I found, namely that the order of `-colors 63 -type Palette` or the other way around matters and one of those is visually worse (my somewhat limited understanding of the operations involved makes it hard to reason or remember which order is better).

Or, perhaps `magick` should be considered a low-level tool and there should be a separate higher-level tool with such niceties, I'm not sure.

**Reply by ilyagr** [reply]
> See also https://imagemagick.org/Usage/basics/#cmdline

In case this is helpful for adjusting the docs, I did find that document on my own, but it only added to my confusion. https://imagemagick.org/Usage/basics/#option_settings seemed to suggest that where I put `-background` wouldn't matter, it does not mention that loading an image is an operator this option can affect. I'm not sure whether I can call it wrong (it's a bit too vague about what an operation is for that), but I did find it misleading.

I've since found other parts of the docs that do state that loading an image is affected by `-background`, but I only found those parts of the docs later. I'm also having trouble finding them a second time; I do find it quite difficult to navigate the docs.

**Comment by snibgo** [comment]
With `magick`, options are executed in the order they are given. A setting such as `-background` affects operations that come *after* it on the command line. In your case, it affects the rasterization of the SVG.

Until a few years ago, in IM v6, options were not necessarily executed in the order they were given.
